File: lib/widgets/website_linking.dart
import 'package:flutter/material.dart';
import '../services/website_verification_service.dart';
import 'simple_website_button.dart';
import 'test_button.dart';

class WebsiteLinking extends StatefulWidget {
  final int userId;

  const WebsiteLinking({
    Key? key,
    required this.userId,
  }) : super(key: key);

  @override
  _WebsiteLinkingState createState() => _WebsiteLinkingState();
}

class _WebsiteLinkingState extends State<WebsiteLinking> {
  final TextEditingController _domainController = TextEditingController();
  List<Map<String, dynamic>> _linkedWebsites = [];
  List<Map<String, dynamic>> _availableWebsites = [];
  bool _isLoadingLinked = false;
  bool _isLoadingAvailable = false;
  bool _isLinking = false;

  @override
  void initState() {
    super.initState();
    _loadLinkedWebsites();
    _loadAvailableWebsites();
  }

  @override
  void dispose() {
    _domainController.dispose();
    super.dispose();
  }

  Future<void> _loadLinkedWebsites() async {
    setState(() => _isLoadingLinked = true);
    
    try {
      final result = await WebsiteVerificationService.getUserWebsites(widget.userId);
      
      if (result['success']) {
        setState(() {
          _linkedWebsites = List<Map<String, dynamic>>.from(result['websites']);
        });
        // Filter available websites
        _filterAvailableWebsites();
      }
    } catch (e) {
      print('Error loading linked websites: $e');
    } finally {
      setState(() => _isLoadingLinked = false);
    }
  }

  Future<void> _loadAvailableWebsites() async {
    setState(() => _isLoadingAvailable = true);
    
    try {
      // You'll need to implement this API endpoint
      final response = await WebsiteVerificationService.getAvailableWebsites();
      
      if (response['success']) {
        setState(() {
          _availableWebsites = List<Map<String, dynamic>>.from(response['websites']);
        });
        _filterAvailableWebsites();
      }
    } catch (e) {
      print('Error loading available websites: $e');
    } finally {
      setState(() => _isLoadingAvailable = false);
    }
  }

  void _filterAvailableWebsites() {
    if (_linkedWebsites.isNotEmpty && _availableWebsites.isNotEmpty) {
      setState(() {
        _availableWebsites = _availableWebsites.where((available) =>
          !_linkedWebsites.any((linked) => linked['website_id'] == available['website_id'])
        ).toList();
      });
    }
  }

  void _onLinkStatusChanged(bool isLinked, Map<String, dynamic>? websiteData) {
    if (isLinked && websiteData != null) {
      setState(() {
        _linkedWebsites.add(websiteData);
        _availableWebsites.removeWhere((website) => 
          website['website_id'] == websiteData['website_id']
        );
      });
    }
  }

  Future<void> _handleLinkNewWebsite() async {
    if (_domainController.text.trim().isEmpty) return;

    setState(() => _isLinking = true);
    
    try {
      final result = await WebsiteVerificationService.verifyAndLinkWebsite(
        domain: _domainController.text.trim(),
        userId: widget.userId,
      );
      
      if (result['success']) {
        setState(() {
          _linkedWebsites.add(result['data']);
          _domainController.clear();
        });
        _filterAvailableWebsites();
        
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(
            content: Text(result['message'] ?? 'Website linked successfully!'),
            backgroundColor: Colors.green,
          ),
        );
      } else {
        if (result['requires_admin'] == true) {
          _showAdminContactDialog(result['admin_last4'] ?? '');
        } else {
          ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(
              content: Text(result['message'] ?? 'Failed to link website'),
              backgroundColor: Colors.red,
            ),
          );
        }
      }
    } catch (e) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: Text('Error: ${e.toString()}'),
          backgroundColor: Colors.red,
        ),
      );
    } finally {
      setState(() => _isLinking = false);
    }
  }

  void _showAdminContactDialog(String last4Digits) {
    showDialog(
      context: context,
      builder: (BuildContext context) {
        return AlertDialog(
          title: const Text('Admin Verification Required'),
          content: Column(
            mainAxisSize: MainAxisSize.min,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              const Text(
                'This website requires admin verification. Please contact the administrator.',
                style: TextStyle(fontSize: 14),
              ),
              const SizedBox(height: 16),
              Container(
                padding: const EdgeInsets.all(12),
                decoration: BoxDecoration(
                  color: Colors.blue.shade50,
                  borderRadius: BorderRadius.circular(8),
                  border: Border.all(color: Colors.blue.shade200),
                ),
                child: Row(
                  children: [
                    Icon(Icons.phone, color: Colors.blue.shade600, size: 20),
                    const SizedBox(width: 8),
                    Text(
                      'Admin: ****$last4Digits',
                      style: TextStyle(
                        fontSize: 16,
                        fontWeight: FontWeight.bold,
                        color: Colors.blue.shade800,
                      ),
                    ),
                  ],
                ),
              ),
            ],
          ),
          actions: [
            TextButton(
              onPressed: () => Navigator.of(context).pop(),
              child: const Text('OK'),
            ),
          ],
        );
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.grey.shade50,
      appBar: AppBar(
        title: const Text('Link Website'),
        backgroundColor: Colors.white,
        foregroundColor: Colors.black,
        elevation: 0,
      ),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            // Link New Website Section
            Container(
              padding: const EdgeInsets.all(16),
              decoration: BoxDecoration(
                color: Colors.white,
                borderRadius: BorderRadius.circular(8),
                border: Border.all(color: Colors.grey.shade200),
              ),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const Text(
                    'Link New Website',
                    style: TextStyle(
                      fontSize: 18,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                  const SizedBox(height: 16),
                  Row(
                    children: [
                      Expanded(
                        child: TextField(
                          controller: _domainController,
                          decoration: InputDecoration(
                            hintText: 'Enter website domain...',
                            border: OutlineInputBorder(
                              borderRadius: BorderRadius.circular(8),
                            ),
                            contentPadding: const EdgeInsets.symmetric(
                              horizontal: 12,
                              vertical: 8,
                            ),
                          ),
                          onSubmitted: (_) => _handleLinkNewWebsite(),
                        ),
                      ),
                      const SizedBox(width: 8),
                      ElevatedButton(
                        onPressed: _isLinking ? null : _handleLinkNewWebsite,
                        style: ElevatedButton.styleFrom(
                          backgroundColor: Colors.blue.shade600,
                          foregroundColor: Colors.white,
                          padding: const EdgeInsets.symmetric(
                            horizontal: 16,
                            vertical: 8,
                          ),
                        ),
                        child: _isLinking
                            ? const SizedBox(
                                width: 16,
                                height: 16,
                                child: CircularProgressIndicator(
                                  strokeWidth: 2,
                                  valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
                                ),
                              )
                            : const Text('Link'),
                      ),
                    ],
                  ),
                ],
              ),
            ),
            
            const SizedBox(height: 24),
            
            // Websites Grid
            Row(
              children: [
                Expanded(
                  child: _buildLinkedWebsitesSection(),
                ),
                const SizedBox(width: 16),
                Expanded(
                  child: _buildAvailableWebsitesSection(),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildLinkedWebsitesSection() {
    return Container(
      padding: const EdgeInsets.all(16),
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(8),
        border: Border.all(color: Colors.grey.shade200),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          const Text(
            'Your Linked Websites',
            style: TextStyle(
              fontSize: 18,
              fontWeight: FontWeight.bold,
            ),
          ),
          const SizedBox(height: 16),
          
          if (_isLoadingLinked)
            const Center(child: CircularProgressIndicator())
          else if (_linkedWebsites.isEmpty)
            Container(
              padding: const EdgeInsets.all(32),
              child: Center(
                child: Column(
                  children: [
                    Icon(
                      Icons.link_off,
                      size: 48,
                      color: Colors.grey.shade300,
                    ),
                    const SizedBox(height: 8),
                    Text(
                      'No linked websites',
                      style: TextStyle(
                        color: Colors.grey.shade500,
                        fontSize: 16,
                      ),
                    ),
                  ],
                ),
              ),
            )
          else
            Column(
              children: _linkedWebsites.map((website) {
                return Container(
                  margin: const EdgeInsets.only(bottom: 12),
                  child: Column(
                    children: [
                      // Website info
                      Container(
                        width: double.infinity,
                        padding: const EdgeInsets.all(12),
                        decoration: BoxDecoration(
                          color: Colors.grey.shade50,
                          borderRadius: BorderRadius.circular(8),
                          border: Border.all(color: Colors.grey.shade200),
                        ),
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            Text(
                              website['website_name'] ?? 'Unknown Website',
                              style: const TextStyle(
                                fontWeight: FontWeight.bold,
                                fontSize: 14,
                              ),
                            ),
                            Text(
                              website['domain'] ?? '',
                              style: TextStyle(
                                color: Colors.grey.shade600,
                                fontSize: 12,
                              ),
                            ),
                          ],
                        ),
                      ),
                      const SizedBox(height: 8),
                      // Unlink button
                      TestButton(
                        text: 'Unlink ${website['website_name'] ?? 'Website'}',
                        onPressed: () async {
                          try {
                            final result = await WebsiteVerificationService.unlinkWebsite(
                              domain: website['domain'] ?? '',
                              userId: widget.userId,
                            );
                            
                            if (result['success']) {
                              setState(() {
                                _linkedWebsites.removeWhere((w) => w['website_id'] == website['website_id']);
                              });
                              _loadAvailableWebsites();
                              
                              ScaffoldMessenger.of(context).showSnackBar(
                                const SnackBar(
                                  content: Text('Website unlinked successfully!'),
                                  backgroundColor: Colors.orange,
                                ),
                              );
                            } else {
                              ScaffoldMessenger.of(context).showSnackBar(
                                SnackBar(
                                  content: Text(result['message'] ?? 'Unlinking failed'),
                                  backgroundColor: Colors.red,
                                ),
                              );
                            }
                          } catch (e) {
                            ScaffoldMessenger.of(context).showSnackBar(
                              SnackBar(
                                content: Text('Error: ${e.toString()}'),
                                backgroundColor: Colors.red,
                              ),
                            );
                          }
                        },
                        color: Colors.red,
                      ),
                    ],
                  ),
                );
              }).toList(),
            ),
        ],
      ),
    );
  }

  Widget _buildAvailableWebsitesSection() {
    return Container(
      padding: const EdgeInsets.all(16),
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(8),
        border: Border.all(color: Colors.grey.shade200),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          const Text(
            'Available Websites',
            style: TextStyle(
              fontSize: 18,
              fontWeight: FontWeight.bold,
            ),
          ),
          const SizedBox(height: 16),
          
          if (_isLoadingAvailable)
            const Center(child: CircularProgressIndicator())
          else if (_availableWebsites.isEmpty)
            Container(
              padding: const EdgeInsets.all(32),
              child: Center(
                child: Column(
                  children: [
                    Icon(
                      Icons.language,
                      size: 48,
                      color: Colors.grey.shade300,
                    ),
                    const SizedBox(height: 8),
                    Text(
                      'No available websites',
                      style: TextStyle(
                        color: Colors.grey.shade500,
                        fontSize: 16,
                      ),
                    ),
                  ],
                ),
              ),
            )
          else
            Column(
              children: _availableWebsites.map((website) {
                return Container(
                  margin: const EdgeInsets.only(bottom: 12),
                  child: Column(
                    children: [
                      // Website info
                      Container(
                        width: double.infinity,
                        padding: const EdgeInsets.all(12),
                        decoration: BoxDecoration(
                          color: Colors.grey.shade50,
                          borderRadius: BorderRadius.circular(8),
                          border: Border.all(color: Colors.grey.shade200),
                        ),
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            Text(
                              website['website_name'] ?? 'Unknown Website',
                              style: const TextStyle(
                                fontWeight: FontWeight.bold,
                                fontSize: 14,
                              ),
                            ),
                            Text(
                              website['domain'] ?? '',
                              style: TextStyle(
                                color: Colors.grey.shade600,
                                fontSize: 12,
                              ),
                            ),
                          ],
                        ),
                      ),
                      const SizedBox(height: 8),
                      // Link button
                      TestButton(
                        text: 'Link ${website['website_name'] ?? 'Website'}',
                        onPressed: () async {
                          try {
                            final result = await WebsiteVerificationService.verifyAndLinkWebsite(
                              domain: website['domain'] ?? '',
                              userId: widget.userId,
                            );
                            
                            if (result['success']) {
                              _onLinkStatusChanged(true, {
                                ...result['data'],
                                'website_id': website['website_id'],
                                'website_name': website['website_name'],
                                'domain': website['domain'],
                              });
                              
                              ScaffoldMessenger.of(context).showSnackBar(
                                const SnackBar(
                                  content: Text('Website linked successfully!'),
                                  backgroundColor: Colors.green,
                                ),
                              );
                            } else {
                              ScaffoldMessenger.of(context).showSnackBar(
                                SnackBar(
                                  content: Text(result['message'] ?? 'Linking failed'),
                                  backgroundColor: Colors.red,
                                ),
                              );
                            }
                          } catch (e) {
                            ScaffoldMessenger.of(context).showSnackBar(
                              SnackBar(
                                content: Text('Error: ${e.toString()}'),
                                backgroundColor: Colors.red,
                              ),
                            );
                          }
                        },
                        color: Colors.blue,
                      ),
                    ],
                  ),
                );
              }).toList(),
            ),
        ],
      ),
    );
  }
}

File: lib/server/components/WebsiteLinking.js
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import WebsiteLinkButton from './WebsiteLinkButton';

const WebsiteLinking = ({ userId }) => {
  const [linkedWebsites, setLinkedWebsites] = useState([]);
  const [availableWebsites, setAvailableWebsites] = useState([]);
  const [isLoadingLinked, setIsLoadingLinked] = useState(false);
  const [isLoadingAvailable, setIsLoadingAvailable] = useState(false);
  const [domain, setDomain] = useState('');

  useEffect(() => {
    loadLinkedWebsites();
    loadAvailableWebsites();
  }, [userId]);

  const loadLinkedWebsites = async () => {
    setIsLoadingLinked(true);
    
    try {
      const response = await axios.get(`/api/websites/user/${userId}`);
      
      if (response.data.success) {
        setLinkedWebsites(response.data.data || []);
      }
    } catch (error) {
      console.error('Error loading linked websites:', error);
    } finally {
      setIsLoadingLinked(false);
    }
  };

  const loadAvailableWebsites = async () => {
    setIsLoadingAvailable(true);
    
    try {
      const response = await axios.get('/api/websites/available');
      
      if (response.data.success) {
        const available = response.data.data || [];
        
        // Filter out already linked websites
        const filteredAvailable = available.filter(website => 
          !linkedWebsites.some(linked => linked.website_id === website.website_id)
        );
        
        setAvailableWebsites(filteredAvailable);
      }
    } catch (error) {
      console.error('Error loading available websites:', error);
    } finally {
      setIsLoadingAvailable(false);
    }
  };

  const onLinkStatusChanged = (isLinked, websiteData) => {
    if (isLinked) {
      // Add to linked websites
      setLinkedWebsites(prev => [...prev, websiteData]);
      
      // Remove from available websites
      setAvailableWebsites(prev => 
        prev.filter(website => website.website_id !== websiteData.website_id)
      );
    }
  };

  const handleUnlinkWebsite = async (websiteId, domain) => {
    try {
      const response = await axios.post('/api/unlink-website', {
        domain: domain,
        user_id: userId,
      });
      
      if (response.data.success) {
        // Remove from linked websites
        setLinkedWebsites(prev => 
          prev.filter(website => website.website_id !== websiteId)
        );
        
        // Add back to available websites
        const unlinkedWebsite = linkedWebsites.find(w => w.website_id === websiteId);
        if (unlinkedWebsite) {
          setAvailableWebsites(prev => [...prev, unlinkedWebsite]);
        }
        
        alert(response.data.message || 'Website unlinked successfully!');
      } else {
        alert(response.data.message || 'Unlinking failed');
      }
    } catch (error) {
      console.error('Error unlinking website:', error);
      alert('Error: ' + (error.response?.data?.message || error.message));
    }
  };

  const handleLinkNewWebsite = async () => {
    if (!domain.trim()) return;
    
    try {
      const response = await axios.post('/api/verify-app', {
        domain: domain.trim(),
        user_id: userId,
      });
      
      if (response.data.success) {
        // Add to linked websites
        setLinkedWebsites(prev => [...prev, response.data.data]);
        
        // Clear domain input
        setDomain('');
        
        // Refresh available websites
        loadAvailableWebsites();
        
        alert('Website linked successfully!');
      } else {
        if (response.data.requires_admin) {
          alert(`Contact Admin: ****${response.data.admin_last4}`);
        } else {
          alert(response.data.message || 'Failed to link website');
        }
      }
    } catch (error) {
      console.error('Error linking website:', error);
      alert('Error: ' + (error.response?.data?.message || error.message));
    }
  };

  return (
    <div className="min-h-screen bg-gray-50 p-4">
      <div className="max-w-6xl mx-auto">
        <h1 className="text-2xl font-bold text-gray-900 mb-6">Link Website</h1>

        {/* Link New Website Section */}
        <div className="bg-white rounded-lg shadow-sm border border-gray-200 p-4 mb-6">
          <h2 className="text-lg font-semibold text-gray-800 mb-4">Link New Website</h2>
          <div className="flex gap-2">
            <input
              type="text"
              value={domain}
              onChange={(e) => setDomain(e.target.value)}
              placeholder="Enter website domain..."
              className="flex-1 px-3 py-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
              onKeyPress={(e) => e.key === 'Enter' && handleLinkNewWebsite()}
            />
            <button
              onClick={handleLinkNewWebsite}
              className="px-4 py-2 bg-blue-600 text-white rounded-md hover:bg-blue-700 transition-colors"
            >
              Link
            </button>
          </div>
        </div>

        <div className="grid grid-cols-1 lg:grid-cols-2 gap-6">
          {/* Your Linked Websites */}
          <div className="bg-white rounded-lg shadow-sm border border-gray-200 p-4">
            <h2 className="text-lg font-semibold text-gray-800 mb-4">Your Linked Websites</h2>
            
            {isLoadingLinked ? (
              <div className="flex justify-center py-8">
                <div className="animate-spin rounded-full h-6 w-6 border-b-2 border-blue-600"></div>
              </div>
            ) : linkedWebsites.length === 0 ? (
              <div className="text-center py-8 text-gray-500">
                <svg className="mx-auto h-12 w-12 text-gray-300 mb-3" fill="currentColor" viewBox="0 0 20 20">
                  <path fillRule="evenodd" d="M12.586 4.586a2 2 0 112.828 2.828l-3 3a2 2 0 01-2.828 0 1 1 0 00-1.414 1.414 4 4 0 005.656 0l3-3a4 4 0 00-5.656-5.656l-1.5 1.5a1 1 0 101.414 1.414l1.5-1.5zm-5 5a2 2 0 012.828 0 1 1 0 101.414-1.414 4 4 0 00-5.656 0l-3 3a4 4 0 105.656 5.656l1.5-1.5a1 1 0 10-1.414-1.414l-1.5 1.5a2 2 0 11-2.828-2.828l3-3z" clipRule="evenodd" />
                </svg>
                <p>No linked websites</p>
              </div>
            ) : (
              <div className="space-y-3">
                {linkedWebsites.map((website) => (
                  <div key={website.website_id} className="flex items-center justify-between p-3 bg-gray-50 rounded-lg">
                    <div className="flex items-center space-x-3">
                      <div className="w-8 h-8 bg-green-100 rounded-full flex items-center justify-center">
                        <svg className="w-4 h-4 text-green-600" fill="currentColor" viewBox="0 0 20 20">
                          <path fillRule="evenodd" d="M12.586 4.586a2 2 0 112.828 2.828l-3 3a2 2 0 01-2.828 0 1 1 0 00-1.414 1.414 4 4 0 005.656 0l3-3a4 4 0 00-5.656-5.656l-1.5 1.5a1 1 0 101.414 1.414l1.5-1.5zm-5 5a2 2 0 012.828 0 1 1 0 101.414-1.414 4 4 0 00-5.656 0l-3 3a4 4 0 105.656 5.656l1.5-1.5a1 1 0 10-1.414-1.414l-1.5 1.5a2 2 0 11-2.828-2.828l3-3z" clipRule="evenodd" />
                        </svg>
                      </div>
                      <div>
                        <h3 className="font-medium text-gray-900">{website.website_name}</h3>
                        <p className="text-sm text-gray-500">{website.domain}</p>
                      </div>
                    </div>
                    <button
                      onClick={() => handleUnlinkWebsite(website.website_id, website.domain)}
                      className="px-3 py-1 text-xs font-medium bg-red-500 text-white rounded-full hover:bg-red-600 transition-colors"
                      title="Unlink this website"
                    >
                      Unlink
                    </button>
                  </div>
                ))}
              </div>
            )}
          </div>

          {/* Available Websites */}
          <div className="bg-white rounded-lg shadow-sm border border-gray-200 p-4">
            <h2 className="text-lg font-semibold text-gray-800 mb-4">Available Websites</h2>
            
            {isLoadingAvailable ? (
              <div className="flex justify-center py-8">
                <div className="animate-spin rounded-full h-6 w-6 border-b-2 border-blue-600"></div>
              </div>
            ) : availableWebsites.length === 0 ? (
              <div className="text-center py-8 text-gray-500">
                <svg className="mx-auto h-12 w-12 text-gray-300 mb-3" fill="currentColor" viewBox="0 0 20 20">
                  <path fillRule="evenodd" d="M4 4a2 2 0 00-2 2v8a2 2 0 002 2h12a2 2 0 002-2V6a2 2 0 00-2-2H4zm12 2H4v8h12V6z" clipRule="evenodd" />
                </svg>
                <p>No available websites</p>
              </div>
            ) : (
              <div className="space-y-3">
                {availableWebsites.map((website) => (
                  <div key={website.website_id} className="flex items-center justify-between p-3 bg-gray-50 rounded-lg">
                    <div className="flex items-center space-x-3">
                      <div className="w-8 h-8 bg-blue-100 rounded-full flex items-center justify-center">
                        <svg className="w-4 h-4 text-blue-600" fill="currentColor" viewBox="0 0 20 20">
                          <path fillRule="evenodd" d="M4 4a2 2 0 00-2 2v8a2 2 0 002 2h12a2 2 0 002-2V6a2 2 0 00-2-2H4zm12 2H4v8h12V6z" clipRule="evenodd" />
                        </svg>
                      </div>
                      <div>
                        <h3 className="font-medium text-gray-900">{website.website_name}</h3>
                        <p className="text-sm text-gray-500">{website.domain}</p>
                      </div>
                    </div>
                    <WebsiteLinkButton
                      domain={website.domain}
                      userId={userId}
                      onLinkStatusChanged={(isLinked, websiteData) => {
                        if (isLinked) {
                          onLinkStatusChanged(true, {
                            ...websiteData,
                            website_id: website.website_id,
                            website_name: website.website_name,
                            domain: website.domain
                          });
                        }
                      }}
                      className="text-sm"
                    />
                  </div>
                ))}
              </div>
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

export default WebsiteLinking;
